### DISCUSSION TITLE
materializing result from EMR

### DISCUSSION THREAD

**Discussion by HPicatto** [dagster-io/dagster#19370] [Q&A] [answered]
Hi, I'm running a job using emr. code block at the end looks like this
```
with open_pipes_session(
            context=context,
            message_reader=self._message_reader,
            context_injector=self._context_injector,
            extras=extras,
        ) as session:
job_flow = self._emr_client.run_job_flow(**emr_job_config)
            get_dagster_logger().debug(f"EMR configuration: {job_flow}")
            try:
                self._emr_client.add_job_flow_steps(
                    JobFlowId=job_flow["JobFlowId"], Steps=[extras.get("step_config")]
                )
                self._poll_till_success(job_flow["JobFlowId"])
            except DagsterExecutionInterruptedError:
                context.log.info("[pipes] execution interrupted, canceling EMR job.")
                self._emr_client.terminate_job_flows(JobFlowIds=[job_flow["JobFlowId"]])
            finally:
                get_dagster_logger("finished")
        return PipesClientCompletedInvocation(session)

```
but I keep getting an error

> dagster._core.errors.DagsterInvariantViolationError: op 'aws_s3__commoncrawl__nodes_only' did not yield or return expected outputs {'result'}. Did you forget to `yield from pipes_session.get_results()` or `return <PipesClient>.run(...).get_results`? If using `open_pipes_session`, `pipes_session.get_results` should be called once after the `open_pipes_session` block has exited to yield any remaining buffered results via `<PipesSession>.get_results()`. If using `<PipesClient>.run`, you should always return `<PipesClient>.run(...).get_results()` or `<PipesClient>.run(...).get_materialize_result()`.
> Stack Trace:
>   File "/home/picatto/.conda/envs/ascii_rs/lib/python3.11/site-packages/dagster/_core/execution/plan/execute_plan.py", line 286, in dagster_event_sequence_for_step
>     for step_event in check.generator(step_events):
>   File "/home/picatto/.conda/envs/ascii_rs/lib/python3.11/site-packages/dagster/_core/execution/plan/execute_step.py", line 478, in core_dagster_event_sequence_for_step
>     for user_event in _step_output_error_checked_user_event_sequence(
>   File "/home/picatto/.conda/envs/ascii_rs/lib/python3.11/site-packages/dagster/_core/execution/plan/execute_step.py", line 160, in _step_output_error_checked_user_event_sequence
>     for user_event in user_event_sequence:
>   File "/home/picatto/.conda/envs/ascii_rs/lib/python3.11/site-packages/dagster/_core/execution/plan/execute_step.py", line 95, in _process_asset_results_to_events
>     for user_event in user_event_sequence:
>   File "/home/picatto/.conda/envs/ascii_rs/lib/python3.11/site-packages/dagster/_core/execution/plan/compute.py", line 260, in execute_core_compute
>     raise DagsterInvariantViolationError(message)

Any idea on how can I avoid this message. I know that on EMR it is processing correctly as the logs over there don't show any problem. and also stores the context correctly and is read by dagster.

**Comment by alangenfeld** [comment]
Did you try to do `yield from session.get_results()` or `return session.get_materialize_result()` instead of `return PipesClientCompletedInvocation(session)` like mentioned in the exception message?

**Reply by HPicatto** [reply]
Yes I did, adding the yield makes that emr process is not even launced. and changing the PipesClientCompletedInvocation for return session.get_materialize_result()  doesn't change the error log

**Reply by alangenfeld** [reply]
Hmm, well my original comment has some slight error in it so hopefully thats the cause. 

Theres some subtlety between what you can call on `session` / `PipesSession` https://docs.dagster.io/_apidocs/pipes#dagster.PipesSession 

and whats on `PipesClientCompletedInvocation` (which needs to get added to the API docs)  https://github.com/dagster-io/dagster/blob/master/python_modules/dagster/dagster/_core/pipes/client.py#L56 

So things I would expect to work are:
* `return PipesClientCompletedInvocation(session).get_materialize_result()`
* `return PipesClientCompletedInvocation(session).get_results()`
* `yield from PipesClientCompletedInvocation(session).get_results()`
* `return session.get_results()`
* `yield from session.get_results()`

**Reply by HPicatto** [reply]
As can be seen on the code I do the `return PipesClientCompletedInvocation(session)` and as soon as I get that object that's where I do `response.get_materialize_result()`. If I do .get_results() I get the same error.
the problem is in the download_messages_chunk on the class below. It yield the last file I tried making intervals even shorter:
while all the files are on yielded, the materialized results figure in the log like this:

> {"__dagster_pipes_version": "0.1", "method": "report_asset_materialization", "params": {"asset_key": "aws_s3/playground_ext/subprocess_asset", "data_version": "alpha", "metadata": {"some_metric": {"raw_value": 4, "type": "int"}, "some_spark_metric": {"raw_value": "tallahassee", "type": "__infer__"}}}}
> {"__dagster_pipes_version": "0.1", "method": "closed", "params": {}}

Probable problem is not on PipesBlobStoreMessageReader but on PipesMessageHandler
```
class PipesS3MessageReader(PipesBlobStoreMessageReader):
    """Message reader that reads messages by periodically reading message chunks from a specified S3
    bucket.

    If `log_readers` is passed, this reader will also start the passed readers
    when the first message is received from the external process.

    Args:
        interval (float): interval in seconds between attempts to download a chunk
        bucket (str): The S3 bucket to read from.
        client (WorkspaceClient): A boto3 client.
        log_readers (Optional[Sequence[PipesLogReader]]): A set of readers for logs on S3.
    """

    def __init__(
        self,
        *,
        interval: float = 10,
        bucket: str,
        key_prefix: str,
        client: client,
        log_readers: Optional[Sequence[PipesLogReader]] = None,
    ):
        super().__init__(
            interval=interval,
            log_readers=log_readers,
        )
        self.key_prefix = key_prefix
        self.bucket = check.str_param(bucket, "bucket")
        self.client = client

    @contextmanager
    def get_params(self) -> Iterator[PipesParams]:
        yield {"bucket": self.bucket, "key_prefix": self.key_prefix}

    def download_messages_chunk(self, index: int, params: PipesParams) -> Optional[str]:
        key = f"{params['key_prefix']}/{index}.json"
        try:
            obj = self.client.get_object(Bucket=self.bucket, Key=key)
            get_dagster_loger().debug(obj["Body"].read().decode("utf-8"))
            return obj["Body"].read().decode("utf-8")
        except ClientError:
            return None

    def no_messages_debug_text(self) -> str:
        return (
            f"Attempted to read messages from S3 bucket {self.bucket}. Expected"
            " PipesS3MessageWriter to be explicitly passed to open_dagster_pipes in the external"
            " process."
        )
```

**Reply by HPicatto** [reply]
I just add a logger to the download_message_chunk in order to check it was receiving the data as it is supposed:
```
 def download_messages_chunk(self, index: int, params: PipesParams) -> Optional[str]:
        key = f"{params['key_prefix']}/{index}.json"
        response = None
        try:
            obj = self.client.get_object(Bucket=self.bucket, Key=key)
            response = obj["Body"].read().decode("utf-8")
            get_dagster_logger().info(response)
        except ClientError:
            get_dagster_logger().debug(f"error reading: {key}")
        return response
```
As can be seen in the pictures below when the message method is "opened" or "log" it works well. But it's not processing when the method is "report_asset_materialization". I'm not seeing any log of an error, so I'm not sure what's going on. The "PipesBlobStoreMessageReader" and "PipesMessageHandler" are the stardard dagster classes and they do work well with other services like Databricks
![open_pipe](https://github.com/dagster-io/dagster/assets/21952067/1bb845e4-562b-468c-b8e3-07d9bfd02b74)
![log_content](https://github.com/dagster-io/dagster/assets/21952067/b936225d-f106-4a5c-a2a9-7ed408c4aca2)
![close_pipe](https://github.com/dagster-io/dagster/assets/21952067/3a1c29eb-56fa-4944-a5c4-e1f57e8f70ab)

**Reply by alangenfeld** [reply]
What version are of Dagster are you using and can you share more of the code in question? Is it available in an open source repo that you could link to?

**Reply by geoHeil** [reply]
I think this is reproducible even for the latest 1.6.1 version.

**Reply by HPicatto** [reply]
[7b07acaf-89e0-4c76-8295-cec0a6d924d9.gz](https://github.com/dagster-io/dagster/files/14066082/7b07acaf-89e0-4c76-8295-cec0a6d924d9.gz)
I'm attaching a log from EMR and a log from DBR for the exact same asset. as can be seen the materialization is the same but it fails for EMR. I'm working on a minimum reproducible example atm
[emr_log.txt](https://github.com/dagster-io/dagster/files/14066012/emr_log.txt)
[dbr_log.txt](https://github.com/dagster-io/dagster/files/14066013/dbr_log.txt)

**Comment by geoHeil** [comment, marked as answer]
clarified in a video call - a return statement was missing
